File: base/Sources/ImportObj.ts

class ImportObj {

	static run = (path: string, replaceExisting = true) => {
		let i = Context.raw.splitBy;
		let isUdim = i == SplitType.SplitUdim;
		let splitCode =
			(i == SplitType.SplitObject || isUdim) ? "o".charCodeAt(0) :
			 i == SplitType.SplitGroup 			   ? "g".charCodeAt(0) :
			 				 			   			 "u".charCodeAt(0); // usemtl

		data_get_blob(path, (b: ArrayBuffer) => {

			if (isUdim) {
				let part = krom_io_obj_parse(b, splitCode, 0, isUdim);
				let name = part.name;
				for (let i = 0; i < part.udims.length; ++i) {
					if (part.udims[i].length == 0) continue;
					let u = i % part.udims_u;
					let v = Math.floor(i / part.udims_u);
					part.name = name + "." + (1000 + v * 10 + u + 1);
					part.inda = part.udims[i];
					i == 0 ? (replaceExisting ? ImportMesh.makeMesh(part, path) : ImportMesh.addMesh(part)) : ImportMesh.addMesh(part);
				}
			}
			else {
				let parts: any[] = [];
				let part = krom_io_obj_parse(b, splitCode, 0, false);
				parts.push(part);
				while (part.has_next) {
					part = krom_io_obj_parse(b, splitCode, part.pos, false);
					// This part does not contain faces (may contain lines only)
					if (part.inda.length == 0) {
						continue;
					}
					parts.push(part);
				}
				if (Context.raw.splitBy == SplitType.SplitMaterial) {
					let posa0;
					let posa1;
					let nora0;
					let nora1;
					let texa0;
					let texa1;
					let inda0;
					let inda1;
					// Merge to single object per material
					for (let i = 0; i < parts.length; ++i) {
						let j = i + 1;
						while (j < parts.length) {
							if (parts[i].name == parts[j].name) {
								posa0 = parts[i].posa;
								posa1 = parts[j].posa;
								nora0 = parts[i].nora;
								nora1 = parts[j].nora;
								texa0 = parts[i].texa != null ? parts[i].texa : null;
								texa1 = parts[j].texa != null ? parts[j].texa : null;
								inda0 = parts[i].inda;
								inda1 = parts[j].inda;
								let voff = Math.floor(posa0.length / 4);
								// Repack merged positions
								let posa32 = new Float32Array(Math.floor(posa0.length / 4) * 3 + Math.floor(posa1.length / 4) * 3);
								for (let k = 0; k < Math.floor(posa0.length / 4); ++k) {
									posa32[k * 3    ] = posa0[k * 4    ] / 32767 * parts[i].scalePos;
									posa32[k * 3 + 1] = posa0[k * 4 + 1] / 32767 * parts[i].scalePos;
									posa32[k * 3 + 2] = posa0[k * 4 + 2] / 32767 * parts[i].scalePos;
								}
								for (let k = 0; k < Math.floor(posa1.length / 4); ++k) {
									posa32[voff * 3 + k * 3    ] = posa1[k * 4    ] / 32767 * parts[j].scalePos;
									posa32[voff * 3 + k * 3 + 1] = posa1[k * 4 + 1] / 32767 * parts[j].scalePos;
									posa32[voff * 3 + k * 3 + 2] = posa1[k * 4 + 2] / 32767 * parts[j].scalePos;
								}
								let scalePos = 0.0;
								for (let k = 0; k < posa32.length; ++k) {
									let f = Math.abs(posa32[k]);
									if (scalePos < f) scalePos = f;
								}
								let inv = 32767 * (1 / scalePos);
								let posa = new Int16Array(posa0.length + posa1.length);
								for (let k = 0; k < Math.floor(posa.length / 4); ++k) {
									posa[k * 4    ] = Math.floor(posa32[k * 3    ] * inv);
									posa[k * 4 + 1] = Math.floor(posa32[k * 3 + 1] * inv);
									posa[k * 4 + 2] = Math.floor(posa32[k * 3 + 2] * inv);
								}
								for (let k = 0; k < Math.floor(posa0.length / 4); ++k) posa[k * 4 + 3] = posa0[k * 4 + 3];
								for (let k = 0; k < Math.floor(posa1.length / 4); ++k) posa[posa0.length + k * 4 + 3] = posa1[k * 4 + 3];
								// Merge normals and uvs
								let nora = new Int16Array(nora0.length + nora1.length);
								let texa = (texa0 != null && texa1 != null) ? new Int16Array(texa0.length + texa1.length) : null;
								let inda = new Uint32Array(inda0.length + inda1.length);
								nora.set(nora0);
								nora.set(nora1, nora0.length);
								if (texa != null) {
									texa.set(texa0);
									texa.set(texa1, texa0.length);
								}
								inda.set(inda0);
								for (let k = 0; k < inda1.length; ++k) inda[k + inda0.length] = inda1[k] + voff;
								parts[i].posa = posa;
								parts[i].nora = nora;
								parts[i].texa = texa;
								parts[i].inda = inda;
								parts[i].scalePos = scalePos;
								parts.splice(j, 1);
							}
							else j++;
						}
					}
				}
				replaceExisting ? ImportMesh.makeMesh(parts[0], path) : ImportMesh.addMesh(parts[0]);
				for (let i = 1; i < parts.length; ++i) {
					ImportMesh.addMesh(parts[i]);
				}
			}
			data_delete_blob(path);
		});
	}
}

File: paint/sources/import_obj.c

#include "global.h"

void import_obj_run(char *path, bool replace_existing) {
	split_type_t i       = context_raw->split_by;
	bool         is_udim = i == SPLIT_TYPE_UDIM;
	i32 split_code = (i == SPLIT_TYPE_OBJECT || is_udim) ? char_code_at("o", 0) : i == SPLIT_TYPE_GROUP ? char_code_at("g", 0) : char_code_at("u", 0); // usemtl

	buffer_t *b = data_get_blob(path);

	if (is_udim) {
		raw_mesh_t *part = obj_parse(b, split_code, 0, is_udim);
		char       *name = part->name;
		for (i32 i = 0; i < part->udims->length; ++i) {
			u32_array_t *a = part->udims->buffer[i];
			if (a->length == 0) {
				continue;
			}
			i32 u      = i % part->udims_u;
			i32 v      = math_floor(i / (float)part->udims_u);
			i32 id     = (1000 + v * 10 + u + 1);
			part->name = string("%s.%s", name, i32_to_string(id));
			part->inda = a;
			if (i == 0) {
				if (replace_existing) {
					import_mesh_make_mesh(part);
				}
				else {
					import_mesh_add_mesh(part);
				}
			}
			else {
				import_mesh_add_mesh(part);
			}
		}
	}
	else {
		raw_mesh_t_array_t *parts = any_array_create_from_raw((void *[]){}, 0);
		raw_mesh_t         *part  = obj_parse(b, split_code, 0, false);
		any_array_push(parts, part);
		while (part->has_next) {
			part = obj_parse(b, split_code, part->pos, false);
			// This part does not contain faces (may contain lines only)
			if (part->inda->length == 0) {
				continue;
			}
			any_array_push(parts, part);
		}
		if (context_raw->split_by == SPLIT_TYPE_MATERIAL) {
			i16_array_t *posa0;
			i16_array_t *posa1;
			i16_array_t *nora0;
			i16_array_t *nora1;
			i16_array_t *texa0;
			i16_array_t *texa1;
			u32_array_t *inda0;
			u32_array_t *inda1;
			// Merge to single object per material
			for (i32 i = 0; i < parts->length; ++i) {
				i32 j = i + 1;
				while (j < parts->length) {
					char *iname = parts->buffer[i]->name;
					char *jname = parts->buffer[j]->name;
					if (string_equals(iname, jname)) {
						posa0    = parts->buffer[i]->posa;
						posa1    = parts->buffer[j]->posa;
						nora0    = parts->buffer[i]->nora;
						nora1    = parts->buffer[j]->nora;
						texa0    = parts->buffer[i]->texa != NULL ? parts->buffer[i]->texa : NULL;
						texa1    = parts->buffer[j]->texa != NULL ? parts->buffer[j]->texa : NULL;
						inda0    = parts->buffer[i]->inda;
						inda1    = parts->buffer[j]->inda;
						i32 voff = math_floor(posa0->length / 4.0);
						// Repack merged positions
						f32_array_t *posa32 = f32_array_create(math_floor(posa0->length / 4.0) * 3 + math_floor(posa1->length / 4.0) * 3);
						for (i32 k = 0; k < math_floor(posa0->length / 4.0); ++k) {
							posa32->buffer[k * 3]     = posa0->buffer[k * 4] / 32767.0 * parts->buffer[i]->scale_pos;
							posa32->buffer[k * 3 + 1] = posa0->buffer[k * 4 + 1] / 32767.0 * parts->buffer[i]->scale_pos;
							posa32->buffer[k * 3 + 2] = posa0->buffer[k * 4 + 2] / 32767.0 * parts->buffer[i]->scale_pos;
						}
						for (i32 k = 0; k < math_floor(posa1->length / 4.0); ++k) {
							posa32->buffer[voff * 3 + k * 3]     = posa1->buffer[k * 4] / 32767.0 * parts->buffer[j]->scale_pos;
							posa32->buffer[voff * 3 + k * 3 + 1] = posa1->buffer[k * 4 + 1] / 32767.0 * parts->buffer[j]->scale_pos;
							posa32->buffer[voff * 3 + k * 3 + 2] = posa1->buffer[k * 4 + 2] / 32767.0 * parts->buffer[j]->scale_pos;
						}
						f32 scale_pos = 0.0;
						for (i32 k = 0; k < posa32->length; ++k) {
							f32 f = math_abs(posa32->buffer[k]);
							if (scale_pos < f) {
								scale_pos = f;
							}
						}
						f32          inv  = 32767 * (1 / (float)scale_pos);
						i16_array_t *posa = i16_array_create(posa0->length + posa1->length);
						for (i32 k = 0; k < math_floor(posa->length / 4.0); ++k) {
							posa->buffer[k * 4]     = math_floor(posa32->buffer[k * 3] * inv);
							posa->buffer[k * 4 + 1] = math_floor(posa32->buffer[k * 3 + 1] * inv);
							posa->buffer[k * 4 + 2] = math_floor(posa32->buffer[k * 3 + 2] * inv);
						}
						for (i32 k = 0; k < math_floor(posa0->length / 4.0); ++k) {
							posa->buffer[k * 4 + 3] = posa0->buffer[k * 4 + 3];
						}
						for (i32 k = 0; k < math_floor(posa1->length / 4.0); ++k) {
							posa->buffer[posa0->length + k * 4 + 3] = posa1->buffer[k * 4 + 3];
						}
						// Merge normals and uvs
						i16_array_t *nora = i16_array_create(nora0->length + nora1->length);
						i16_array_t *texa = (texa0 != NULL && texa1 != NULL) ? i16_array_create(texa0->length + texa1->length) : NULL;
						u32_array_t *inda = u32_array_create(inda0->length + inda1->length);

						for (i32 k = 0; k < nora0->length; ++k) {
							nora->buffer[k] = nora0->buffer[k];
						}
						for (i32 k = 0; k < nora1->length; ++k) {
							nora->buffer[k + nora0->length] = nora1->buffer[k];
						}

						if (texa != NULL) {
							for (i32 k = 0; k < texa0->length; ++k) {
								texa->buffer[k] = texa0->buffer[k];
							}
							for (i32 k = 0; k < texa1->length; ++k) {
								texa->buffer[k + texa0->length] = texa1->buffer[k];
							}
						}

						for (i32 k = 0; k < inda0->length; ++k) {
							inda->buffer[k] = inda0->buffer[k];
						}
						for (i32 k = 0; k < inda1->length; ++k) {
							inda->buffer[k + inda0->length] = inda1->buffer[k] + voff;
						}

						parts->buffer[i]->posa      = posa;
						parts->buffer[i]->nora      = nora;
						parts->buffer[i]->texa      = texa;
						parts->buffer[i]->inda      = inda;
						parts->buffer[i]->scale_pos = scale_pos;
						array_splice(parts, j, 1);
					}
					else {
						j++;
					}
				}
			}
		}

		replace_existing ? import_mesh_make_mesh(parts->buffer[0]) : import_mesh_add_mesh(parts->buffer[0]);
		for (i32 i = 1; i < parts->length; ++i) {
			import_mesh_add_mesh(parts->buffer[i]);
		}
	}
	data_delete_blob(path);
}
